File: Games/Uzta/src/main/java/jokatu/game/games/uzta/player/UztaPlayer.java
package jokatu.game.games.uzta.player;

import jokatu.game.event.GameEvent;
import jokatu.game.games.uzta.event.ResourcesUpdatedEvent;
import jokatu.game.games.uzta.game.UztaColour;
import jokatu.game.games.uzta.graph.NodeType;
import jokatu.game.player.Player;
import ophelia.collections.bag.BagUtils;
import ophelia.collections.bag.BaseIntegerBag;
import ophelia.collections.bag.HashBag;
import ophelia.collections.bag.ModifiableIntegerBag;
import ophelia.event.observable.AbstractSynchronousObservable;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.text.MessageFormat;

import static java.util.stream.Collectors.joining;

public class UztaPlayer extends AbstractSynchronousObservable<GameEvent> implements Player {

	private final String name;
	private UztaColour colour;

	private ModifiableIntegerBag<NodeType> resources = new HashBag<>();

	public UztaPlayer(@NotNull String username) {
		this.name = username;
	}

	public void setColour(@Nullable UztaColour colour) {
		this.colour = colour;
	}

	@Nullable
	public UztaColour getColour() {
		return colour;
	}

	public void giveResources(@NotNull BaseIntegerBag<NodeType> givenResources) {
		if (resources.getSum(givenResources).hasLackingItems()) {
			throw new NotEnoughResourcesException(
					"{0} does not have enough resources {1}",
					name,
					BagUtils.presentBag(
							givenResources,
							NodeType::getNumber,
							joining(", ", "to be given ", " "),
							joining(", ", "and to give ", ".")
					)
			);
		}
		resources.merge(givenResources);
		fireEvent(new ResourcesUpdatedEvent(this, givenResources));
	}

	@NotNull
	public BaseIntegerBag<NodeType> getResourcesLeftAfter(@NotNull BaseIntegerBag<NodeType> requiredResources) {
		return resources.getDifference(requiredResources);
	}

	@NotNull
	@Override
	public String getName() {
		return name;
	}

	@Override
	public String toString() {
		return name;
	}

	public int getNumberOfType(@NotNull NodeType type) {
		return resources.getNumberOf(type);
	}

	public static class NotEnoughResourcesException extends RuntimeException {
		public NotEnoughResourcesException(@NotNull String pattern, Object... arguments) {
			super(MessageFormat.format(pattern, arguments));
		}
	}
}

File: Games/Uzta/src/main/kotlin/jokatu/game/games/uzta/player/UztaPlayer.kt
package jokatu.game.games.uzta.player

import jokatu.game.event.GameEvent
import jokatu.game.games.uzta.event.ResourcesUpdatedEvent
import jokatu.game.games.uzta.game.UztaColour
import jokatu.game.games.uzta.graph.NodeType
import jokatu.game.player.Player
import ophelia.collections.bag.BagUtils
import ophelia.collections.bag.BaseIntegerBag
import ophelia.collections.bag.HashBag
import ophelia.event.observable.AbstractSynchronousObservable
import java.util.stream.Collectors.joining

class UztaPlayer(override val name: String) : AbstractSynchronousObservable<GameEvent>(), Player {
	var colour: UztaColour? = null

	private val resources = HashBag<NodeType>()

	fun giveResources(givenResources: BaseIntegerBag<NodeType>) {
		if (resources.getSum(givenResources).hasLackingItems()) {
			val bagMessage = BagUtils.presentBag(
					givenResources,
					{ obj, number -> obj.getNumber(number) },
					joining(", ", "to be given ", " "),
					joining(", ", "and to give ", ".")
			)
			throw NotEnoughResourcesException("$name does not have enough resources $bagMessage")
		}
		resources.merge(givenResources)
		fireEvent(ResourcesUpdatedEvent(this, givenResources))
	}

	fun getResourcesLeftAfter(requiredResources: BaseIntegerBag<NodeType>): BaseIntegerBag<NodeType> {
		return resources.getDifference(requiredResources)
	}

	override fun toString(): String {
		return name
	}

	fun getNumberOfType(type: NodeType): Int {
		return resources.getNumberOf(type)
	}

	fun getResources(): BaseIntegerBag<NodeType> {
		return resources
	}

	class NotEnoughResourcesException(message: String) : RuntimeException(message)
}
